WIP: 1.0.0: Schema update

betsybookwyrm opened the pull request:
Most notably adds the ability to add multiple files of JSON results to the same database.

To do list:

- [ ]  multifile test with an additional file that can actually be added to the git repository
- [ ]  update the media table (decision - should it allow multiple listings per media key over time or just one?)
- [x]  put the only-the-latest tweet and user views back in

betsybookwyrm committed:
Add data model diagram

These are rough ER diagrams for the abstract schema / data model of the
database.

betsybookwyrm committed:
Split multisource tables into tweet_ and user_ tables

Addresses the structural part of  #12

betsybookwyrm committed:
Rename and normalise hashtag columns

As suggested by @SamHames in #12

- Renames *_hashtag.tag to *_hashtag.hashtag for clarity
- Adds *_hashtag.hashtag_lower column for normalised text as twitter's
  hashtags are case insensitive.

betsybookwyrm committed:
Update schema docs

betsybookwyrm committed:
Make entity tables unique per tweet/user

Adds a primary key to each of the entity tables:
- *_url: (tweet/user_id, url)
- *_hashtag: (tweet/user_id, hashtag)
- *_mention: (tweet/user_id, username)

This has the consequence of making each pair unique. For example,
if a tweet mentions the same user twice, there will only be one row
for that tweet_id with that username in the tweet_mention table.

betsybookwyrm committed:
Handle primary key table constraints when parsing SQL

betsybookwyrm committed:
Multifile granularity

Saves page metadata as loaded and tracks which tweet/user came from
which page

betsybookwyrm committed:
addendum

betsybookwyrm committed:
lint

betsybookwyrm committed:
Update schema diagram, and add link in readme

betsybookwyrm committed:
Add setuptools_scm to dev requirements

betsybookwyrm committed:
Add newer versions of python to test setup

betsybookwyrm committed:
Lint!

betsybookwyrm committed:
Seems github actions can't handle '3.10' as python version

betsybookwyrm commented:
I've released the current state (tag 1.0.0a1) to PyPI as an alpha so people can work with it and see how it feels. You'll need to use the `--pre` tag in your pip commands to install the latest pre-release, otherwise it will only install v0.4.

```shell
pip install --upgrade --pre tidy_tweet
```

betsybookwyrm committed:
Add views back in

betsybookwyrm commented:
Ok, sorted out the tweet and user tables so that they're now the views that only have the last-retrieved results for each tweet/user. Released as alpha 2 - https://github.com/QUT-Digital-Observatory/tidy_tweet/releases/tag/1.0.0a2

Quiet27 commented:
Thanks for adding the views. I'm working my way through updating the queries in tweetexploR and came across an issue. I previously had this query:

```
SELECT count(*) as `tweet_count`, username
FROM tweet
LEFT JOIN (
	SELECT username, id
	FROM user ) user
ON user.id = tweet.author_id
GROUP BY username
ORDER BY tweet_count DESC;
```

which I have modified as follows:

```
SELECT count(*) as `tweet_count`, username
FROM tweet
  LEFT JOIN (
    SELECT username, id
    FROM user ) user
  ON user.id = tweet.author_id
GROUP BY username
ORDER BY tweet_count DESC;
```

but I am getting an error "SQL error or missing database (ambiguous column name: id)". I also get the same error when trying to open the `user` view in DBeaver.

betsybookwyrm committed:
fix user view

betsybookwyrm committed:
Tweak column types and use strict typing

A few issues were encountered with the schema when playing with loading
tidy_tweet data into postgres. These type changes clear those up.

- media.alt_text was incorrectly of type string (which doesn't exist)
- results_page.result_count was text and is now number
- tables are now created in strict mode to help identify such issues
  in future

betsybookwyrm committed:
Add error type with metadata for page parsing failures

betsybookwyrm committed:
Add a flag to determine whether to create tables in strict mode

Some SQLite tools have trouble with tables that were created in strict mode.

betsybookwyrm committed:
Schema change: (page, file) now page key to enable db merging

betsybookwyrm committed:
Expose option to skip strict mode in table creation

betsybookwyrm committed:
Update schema diagram
